util/label/label.rs

```diff
@@ -9,49 +9,144 @@ use label_error::LabelError;
 
 impl<'s> Label<'s> {
-    fn new(
-        repository_name: Option<&'s str>,
-        package_name: Option<&'s str>,
-        name: &'s str,
-    ) -> Label<'s> {
-        Label {
-            repository_name,
+    /// Parse and analyze given str.
+    pub fn analyze(input: &'s str) -> Result<Label<'s>> {
+        let label = input;
+
+        if label.is_empty() {
+            return Err(LabelError(err(
+                label,
+                "Empty string cannot be parsed into a label.",
+            )));
+        }
+
+        if label.starts_with(':') {
+            return match consume_name(input, label)? {
+                None => Err(LabelError(err(
+                    label,
+                    "Relative packages must have a name.",
+                ))),
+                Some(name) => Ok(Label::Relative { target_name: name }),
+            };
+        }
+
+        let (input, repository) = consume_repository_name(input, label)?;
+
+        // Shorthand labels such as `@repo` are expanded to `@repo//:repo`.
+        if input.is_empty() {
+            if let Some(ref repo) = repository {
+                let target_name = repo.repo_name();
+                if target_name.is_empty() {
+                    return Err(LabelError(err(
+                        label,
+                        "invalid target name: empty target name",
+                    )));
+                } else {
+                    return Ok(Label::Absolute {
+                        repository,
+                        package_name: None,
+                        target_name,
+                    });
+                };
+            }
+        }
+        let (input, package_name) = consume_package_name(input, label)?;
+        let name = consume_name(input, label)?;
+        let name = match (package_name, name) {
+            (None, None) => {
+                return Err(LabelError(err(
+                    label,
+                    "labels must have a package and/or a name.",
+                )))
+            }
+            (Some(package_name), None) => name_from_package(package_name),
+            (_, Some(name)) => name,
+        };
+
+        Ok(Label::Absolute {
+            repository,
             package_name,
-            name,
+            target_name: name,
+        })
+    }
+
+    pub fn is_relative(&self) -> bool {
+        match self {
+            Label::Absolute { .. } => false,
+            Label::Relative { .. } => true,
+        }
+    }
+
+    pub fn repo(&self) -> Option<&Repository<'s>> {
+        match self {
+            Label::Absolute { repository, .. } => repository.as_ref(),
+            Label::Relative { .. } => None,
+        }
+    }
+
+    pub fn repo_name(&self) -> Option<&'s str> {
+        match self {
+            Label::Absolute { repository, .. } => repository.as_ref().map(|repo| repo.repo_name()),
+            Label::Relative { .. } => None,
         }
     }
 
     pub fn packages(&self) -> Vec<&'s str> {
```

Comment:
I'm not sure what this function exists for - it seems unused, and its semantics seem generally unhelpful? Should we delete it?

Reply:
It’s mostly around because it was already there. But yeah, I think the semantics around “packages” is not knowable by just the label. If a package is a directory containing a BUILD file then you can have a directory with another BUILD file multiple directories deep. I think it’d be better to have users accept the semantics applied here within their own repo. Simple enough to split the package in user code